### PR TITLE
GOBBLIN-957: Add recursion eliminating converter for Avro

### PR DESCRIPTION
**Pull request by shirshanka**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-957

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Implements a simple recursion elimination solution that just drops fields that introduce recursion in the schema.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
  Added tests in AvroUtilsTest

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-core/src/main/java/org/apache/gobblin/converter/avro/AvroRecursionEliminatingConverter.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+public class AvroRecursionEliminatingConverter extends AvroToAvroConverterBase {
+  @Override
+  public Schema convertSchema(Schema inputSchema, WorkUnitState workUnit)
+      throws SchemaConversionException {
+    Pair<Schema, List<AvroUtils.SchemaEntry>> results = AvroUtils.dropRecursiveFields(inputSchema);
+    List<AvroUtils.SchemaEntry> recursiveFields = results.getSecond();
+    if (!recursiveFields.isEmpty()) {
+      log.info("Schema " + inputSchema.getFullName() + " is recursive. Will drop fields [" +
+          recursiveFields.stream().map(entry -> entry.getFieldName()).collect(Collectors.joining(",")) + "]");
+      log.info("Schema %s is recursive. Will drop fields [%s]", inputSchema.getFullName(),
```
Comment (autumnust) [inline gobblin-core/src/main/java/org/apache/gobblin/converter/avro/AvroRecursionEliminatingConverter.java]: Why this is being logged twice?

File: gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java [hunk 2]
```diff
@@ -935,4 +941,141 @@ public static Schema overrideNameAndNamespace(Schema input, String nameOverride,
...
+          return copiedMapSchema;
+        }
+      }
+      default: {
+        return schema;
+      }
+    }
+  }
+
+  private static void copyFieldProperties(Schema.Field field, Schema.Field copiedField) {
```
Comment (autumnust) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java]: What about other elements of `Schema` object like `doc` ?

File: gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java [hunk 3]
```diff
@@ -935,4 +941,141 @@ public static Schema overrideNameAndNamespace(Schema input, String nameOverride,
...
+   *    "fields": [
+   *    {"name": "name", "type": "string"}
+   *    ]
+   * }
+   *
+   * @param schema
+   * @return a Pair of (The transformed schema with recursion eliminated, A list of @link{SchemaEntry} objects which
+   * represent the fields that were removed from the original schema)
+   */
+  public static Pair<Schema, List<SchemaEntry>> dropRecursiveFields(Schema schema) {
```
Comment (autumnust) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java]: Can we add one more utility method like `containsRecurisveSchema(Schema schema)` as a easy-to-check tool ?

File: gobblin-core/src/main/java/org/apache/gobblin/converter/avro/AvroRecursionEliminatingConverter.java (hunk 1 above)
Comment (shirshanka) [reply gobblin-core/src/main/java/org/apache/gobblin/converter/avro/AvroRecursionEliminatingConverter.java]: fixed

File: gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java (hunk 2 above)
Comment (shirshanka) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java]: those are not copyable... they need to be provided when you create the field. 
You'll see that in the code above where I do: 

Schema.Field copiedField =
                new Schema.Field(field.name(), copiedFieldSchema, field.doc(), field.defaultValue(), field.order());

Will add javadoc on this method to make this clearer.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java (hunk 3 above)
Comment (shirshanka) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java]: added

**Pull request closed without merging**